gr-azure-software-radio/lib/difi_source_cpp_impl.cc

```diff
@@ -84,24 +112,95 @@ namespace gr {
                               gr_vector_const_void_star &input_items,
                               gr_vector_void_star &output_items)
     {
+      // check the file descriptor for incoming client connection
+      if(d_socket_type == SOCK_STREAM)
+      {
+        struct timeval tv;
+        tv.tv_sec = 0;
+        tv.tv_usec = 10000;
+
+        fd_set read_fds = d_rset;
+        int fd_max = *std::max_element(d_fds.begin(), d_fds.end());
+
+        if (select(fd_max+1, &read_fds, NULL, NULL, &tv) < 0)
+        {
+          GR_LOG_ERROR(this->d_logger, "Could not select server fd");
+          throw std::runtime_error("Could not select server fd");
+        }
+
+        for(auto fd : d_fds)
+        {
+          if(FD_ISSET(fd, &read_fds))
+          {
+            if(fd == d_socket)
+            {
+              socklen_t addrlen = sizeof(d_servaddr);
+              d_client_socket = accept(d_socket, (struct sockaddr*)&d_servaddr, &addrlen);
+              
+              if (d_client_socket < 0)
+              {
+                GR_LOG_ERROR(this->d_logger, "Could not accept incoming connection");
+                throw std::runtime_error("Could not accept incoming connection");
+              }
+
+              FD_SET(d_client_socket, &d_rset);
+              d_fds.push_back(d_client_socket);
+            }
+            else
+            {
+              T *out = reinterpret_cast<T *>(output_items[0]);
+              return buffer_and_send(out, noutput_items);
+            }
+          }
+        }
+      }
+      else
+      {
+        T *out = reinterpret_cast<T *>(output_items[0]);
+        return buffer_and_send(out, noutput_items);
+      }
 
-      T *out = reinterpret_cast<T *>(output_items[0]);
-      return buffer_and_send(out, noutput_items);
+      return 0;
     }
 
     template <class T>
     int difi_source_cpp_impl<T>::buffer_and_send(T *out, int noutput_items)
     {
       boost::this_thread::disable_interruption disable_interrupt;
       socklen_t len = sizeof(d_servaddr);
-      int size_gotten = recvfrom(d_socket, &d_buffer[0], d_buffer.size(),
-                                 MSG_WAITALL, (struct sockaddr *)&d_servaddr, &len);
-
+      int size_gotten = -1;
+      if(d_socket_type == SOCK_DGRAM)
+      {
+        size_gotten = recvfrom(d_socket, &d_packet_buffer[0], d_packet_buffer.size(),
+                                  MSG_WAITALL, (struct sockaddr *)&d_servaddr, &len);
+      }
+      else
+      {
+          // compute the number of bytes per packet
+          // per DIFI version 1.0.0 spec each packet contains the packet size (number of 32-bit words) in the header
+          // Packet Size: Bits 0-15
+          size_gotten = recv(d_client_socket, &d_packet_buffer[0], sizeof(int32_t),MSG_WAITALL);
+          // client socket closed, remove from fd list
+          if(size_gotten <= 0)
+          {
+            reset_tcp_connection();
```

Comment:
do we want to close? What if it is a burst connection? Like just bursting 10000 samples ever 1 second?